### ISSUE TITLE
ASE calculator results in ImportError

### ISSUE THREAD

**Issue by Andrew-S-Rosen** [tblite/tblite#24]
I installed `tblite` by `git clone`'ing the repo and then doing `export CC=gcc`, `pip install .`, and `pip install '.[ase]` as mentioned in the docs.

Then, doing 
```python
from tblite.ase import TBLite
```
results in
```python
Traceback (most recent call last):
  File "/home/arosen/software/tblite/python/tblite/library.py", line 27, in <module>
    from ._libtblite import ffi, lib
ModuleNotFoundError: No module named 'tblite._libtblite'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/arosen/software/tblite/python/tblite/ase.py", line 32, in <module>
    from .interface import Calculator
  File "/home/arosen/software/tblite/python/tblite/interface.py", line 26, in <module>
    from . import library
  File "/home/arosen/software/tblite/python/tblite/library.py", line 29, in <module>
    raise ImportError("tblite C extension unimportable, cannot use C-API")
ImportError: tblite C extension unimportable, cannot use C-API
```
 Using `conda install tblite` doesn't help either.

**Comment by awvwgk** [comment]
Are you running out of source? In this case make sure the file `python/tblite/_libtblite.abi3.so` has actually been generated by `setup.py`. In `pip install -vv` you should find the `build_ext` hook running as shown below:

```
    running build_ext
    generating cffi module 'build/temp.linux-x86_64-3.9/tblite._libtblite.c'
    creating build
    creating build/temp.linux-x86_64-3.9
    building 'tblite._libtblite' extension
    creating build/temp.linux-x86_64-3.9/build
    creating build/temp.linux-x86_64-3.9/build/temp.linux-x86_64-3.9
    gcc -pthread -B /home/awvwgk/software/opt/conda/envs/cffi/compiler_compat -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /home/awvwgk/software/opt/conda/envs/cffi/include -fPIC -O2 -isystem /home/awvwgk/software/opt/conda/envs/cffi/include -fPIC -I/home/awvwgk/projects/src/git/mctc-tools/tblite/python/../_dist/include -I/home/awvwgk/software/opt/conda/envs/cffi/include/python3.9 -c build/temp.linux-x86_64-3.9/tblite._libtblite.c -o build/temp.linux-x86_64-3.9/build/temp.linux-x86_64-3.9/tblite._libtblite.o
    creating build/lib.linux-x86_64-3.9
    creating build/lib.linux-x86_64-3.9/tblite
    gcc -pthread -B /home/awvwgk/software/opt/conda/envs/cffi/compiler_compat -shared -Wl,-rpath,/home/awvwgk/software/opt/conda/envs/cffi/lib -Wl,-rpath-link,/home/awvwgk/software/opt/conda/envs/cffi/lib -L/home/awvwgk/software/opt/conda/envs/cffi/lib -Wl,-rpath,/home/awvwgk/software/opt/conda/envs/cffi/lib -Wl,-rpath-link,/home/awvwgk/software/opt/conda/envs/cffi/lib -L/home/awvwgk/software/opt/conda/envs/cffi/lib build/temp.linux-x86_64-3.9/build/temp.linux-x86_64-3.9/tblite._libtblite.o -L/home/awvwgk/projects/src/git/mctc-tools/tblite/python/../_dist/lib -Wl,-R/home/awvwgk/projects/src/git/mctc-tools/tblite/python/../_dist/lib -ltblite -o build/lib.linux-x86_64-3.9/tblite/_libtblite.abi3.so
    copying build/lib.linux-x86_64-3.9/tblite/_libtblite.abi3.so -> tblite
```

**Comment by Andrew-S-Rosen** [comment]
Hi @awvwgk. The suggested `.so` file is not found in the directory, so I don't think it has been generated during the `pip install` process.

When I `pip install`, I get the text below. This is all being done from the `tblite/python` folder.
```python
Using pip 22.0.4 from /home/arosen/anaconda3/envs/py38/lib/python3.8/site-packages/pip (python 3.8)
Non-user install because site-packages writeable
Created temporary directory: /tmp/pip-ephem-wheel-cache-dswxtj55
Created temporary directory: /tmp/pip-req-tracker-rg4l2_41
Initialized build tracking at /tmp/pip-req-tracker-rg4l2_41
Created build tracker: /tmp/pip-req-tracker-rg4l2_41
Entered build tracker: /tmp/pip-req-tracker-rg4l2_41
Created temporary directory: /tmp/pip-install-h64aw28g
Processing /home/arosen/tblite/python
  Added file:///home/arosen/tblite/python to build tracker '/tmp/pip-req-tracker-rg4l2_41'
  Running setup.py (path:/home/arosen/tblite/python/setup.py) egg_info for package from file:///home/arosen/tblite/python
  Created temporary directory: /tmp/pip-pip-egg-info-dgsdp3zk
  Running command python setup.py egg_info
  running egg_info
  creating /tmp/pip-pip-egg-info-dgsdp3zk/tblite.egg-info
  writing /tmp/pip-pip-egg-info-dgsdp3zk/tblite.egg-info/PKG-INFO
  writing dependency_links to /tmp/pip-pip-egg-info-dgsdp3zk/tblite.egg-info/dependency_links.txt
  writing requirements to /tmp/pip-pip-egg-info-dgsdp3zk/tblite.egg-info/requires.txt
  writing top-level names to /tmp/pip-pip-egg-info-dgsdp3zk/tblite.egg-info/top_level.txt
  writing manifest file '/tmp/pip-pip-egg-info-dgsdp3zk/tblite.egg-info/SOURCES.txt'
  reading manifest file '/tmp/pip-pip-egg-info-dgsdp3zk/tblite.egg-info/SOURCES.txt'
  writing manifest file '/tmp/pip-pip-egg-info-dgsdp3zk/tblite.egg-info/SOURCES.txt'
  Preparing metadata (setup.py) ... done
  Source in /home/arosen/tblite/python has version 0.0.0, which satisfies requirement tblite==0.0.0 from file:///home/arosen/tblite/python
  Removed tblite==0.0.0 from file:///home/arosen/tblite/python from build tracker '/tmp/pip-req-tracker-rg4l2_41'
Requirement already satisfied: cffi in /home/arosen/anaconda3/envs/py38/lib/python3.8/site-packages (from tblite==0.0.0) (1.15.0)
Requirement already satisfied: numpy in /home/arosen/anaconda3/envs/py38/lib/python3.8/site-packages (from tblite==0.0.0) (1.22.3)
Requirement already satisfied: pycparser in /home/arosen/anaconda3/envs/py38/lib/python3.8/site-packages (from cffi->tblite==0.0.0) (2.21)
Created temporary directory: /tmp/pip-unpack-3y81xa0y
Building wheels for collected packages: tblite
  Created temporary directory: /tmp/pip-wheel-_bslftiv
  Building wheel for tblite (setup.py) ...   Destination directory: /tmp/pip-wheel-_bslftiv
  Running command python setup.py bdist_wheel
  running bdist_wheel
  running build
  running build_py
  running build_ext
  generating cffi module 'build/temp.linux-x86_64-cpython-38/tblite._libtblite.c'
  already up-to-date
  /home/arosen/anaconda3/envs/py38/lib/python3.8/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
    warnings.warn(
  installing to build/bdist.linux-x86_64/wheel
  running install
  running install_lib
  creating build/bdist.linux-x86_64/wheel
  creating build/bdist.linux-x86_64/wheel/tblite
  copying build/lib.linux-x86_64-cpython-38/tblite/_libtblite.abi3.so -> build/bdist.linux-x86_64/wheel/tblite
  copying build/lib.linux-x86_64-cpython-38/tblite/interface.py -> build/bdist.linux-x86_64/wheel/tblite
  copying build/lib.linux-x86_64-cpython-38/tblite/test_ase.py -> build/bdist.linux-x86_64/wheel/tblite
  copying build/lib.linux-x86_64-cpython-38/tblite/ase.py -> build/bdist.linux-x86_64/wheel/tblite
  copying build/lib.linux-x86_64-cpython-38/tblite/library.py -> build/bdist.linux-x86_64/wheel/tblite
  copying build/lib.linux-x86_64-cpython-38/tblite/test_library.py -> build/bdist.linux-x86_64/wheel/tblite
  copying build/lib.linux-x86_64-cpython-38/tblite/__init__.py -> build/bdist.linux-x86_64/wheel/tblite
  copying build/lib.linux-x86_64-cpython-38/tblite/test_interface.py -> build/bdist.linux-x86_64/wheel/tblite
  running install_egg_info
  running egg_info
  writing tblite.egg-info/PKG-INFO
  writing dependency_links to tblite.egg-info/dependency_links.txt
  writing requirements to tblite.egg-info/requires.txt
  writing top-level names to tblite.egg-info/top_level.txt
  reading manifest file 'tblite.egg-info/SOURCES.txt'
  writing manifest file 'tblite.egg-info/SOURCES.txt'
  Copying tblite.egg-info to build/bdist.linux-x86_64/wheel/tblite-0.0.0-py3.8.egg-info
  running install_scripts
  creating build/bdist.linux-x86_64/wheel/tblite-0.0.0.dist-info/WHEEL
  creating '/tmp/pip-wheel-_bslftiv/tblite-0.0.0-cp38-cp38-linux_x86_64.whl' and adding 'build/bdist.linux-x86_64/wheel' to it
  adding 'tblite/__init__.py'
  adding 'tblite/_libtblite.abi3.so'
  adding 'tblite/ase.py'
  adding 'tblite/interface.py'
  adding 'tblite/library.py'
  adding 'tblite/test_ase.py'
  adding 'tblite/test_interface.py'
  adding 'tblite/test_library.py'
  adding 'tblite-0.0.0.dist-info/METADATA'
  adding 'tblite-0.0.0.dist-info/WHEEL'
  adding 'tblite-0.0.0.dist-info/top_level.txt'
  adding 'tblite-0.0.0.dist-info/RECORD'
  removing build/bdist.linux-x86_64/wheel
done
  Created wheel for tblite: filename=tblite-0.0.0-cp38-cp38-linux_x86_64.whl size=40680 sha256=cf097672611c0292dbd4a8506c9a293ba31c988c7019151b5c0178d8e4387c2a
  Stored in directory: /tmp/pip-ephem-wheel-cache-dswxtj55/wheels/4e/33/49/323d3e7bfef6510f60266882e541b869ae06c11a82d2694c0e
Successfully built tblite
Installing collected packages: tblite

Successfully installed tblite-0.0.0
Removed build tracker: '/tmp/pip-req-tracker-rg4l2_41'
```

**Comment by awvwgk** [comment]
>```
>...
>  copying build/lib.linux-x86_64-cpython-38/tblite/_libtblite.abi3.so -> build/bdist.linux-x86_64/wheel/tblite
>...
>  adding 'tblite/_libtblite.abi3.so'
>...
>```

Well the extension module is built, maybe try to import from a different directory to avoid having the source directory added implicitly to your `PYTHONPATH` or, if you want to run out-of-source, install `tblite` in development mode (`-e`).

**Comment by Andrew-S-Rosen** [comment]
I tried installing `tblite` in development mode now, and this is what I get back in return.
```python
----> 1 from tblite.ase import TBLite

File ~/tblite/python/tblite/ase.py:32, in <module>
     27     raise ModuleNotFoundError("This submodule requires ASE installed")
     30 from typing import List, Optional
---> 32 from .interface import Calculator
     33 import ase.calculators.calculator
     34 from ase.atoms import Atoms

File ~/tblite/python/tblite/interface.py:26, in <module>
     23 import numpy as np
     24 from typing import Optional
---> 26 from . import library
     29 class Structure:
     30     """
     31     .. Molecular structure data
     32
   (...)
     56         on invalid input, like incorrect shape / type of the passed arrays
     57     """

File ~/tblite/python/tblite/library.py:78, in <module>
     72         return value
     74     return handle_error
     77 @ffi.def_extern()
---> 78 def logger_callback(message, nchar, data):
     79     """Custom logger callback to write output in a Python friendly way"""
     81     print(ffi.unpack(message, nchar).decode())

error: ffi.def_extern('logger_callback'): no 'extern "Python"' function with this name
```
Based on the error, I then tried running the import from the `tblite/python/tblite` directory and got back the following.
```python
----> 1 from tblite.ase import TBLite

File ~/tblite/python/tblite/ase.py:25, in <module>
     16 """
     17 The Python API of *tblite* natively support integration with the atomic simulation environment (`ASE`_).
     18 By constructing a calculator most functionality of ASE is readily available.
   (...)
     21 .. _ase: https://wiki.fysik.dtu.dk/ase/
     22 """
     24 try:
---> 25     import ase
     26 except ModuleNotFoundError:
     27     raise ModuleNotFoundError("This submodule requires ASE installed")

File ~/tblite/python/tblite/ase.py:32, in <module>
     27     raise ModuleNotFoundError("This submodule requires ASE installed")
     30 from typing import List, Optional
---> 32 from .interface import Calculator
     33 import ase.calculators.calculator
     34 from ase.atoms import Atoms

ImportError: attempted relative import with no known parent package
```
I do see the `_libtblite.abi3.so` file though!

**Comment by awvwgk** [comment]
Which version of tblite are you compiling the extension module against? If you have installed `tblite` from conda-forge (version 0.1.0) make sure to also checkout the repository at the `v0.1.0` tag. If you use the latest version, you also have to build the `tblite` from source since I added a breaking API change for the upcoming version 0.2.0.

**Comment by Andrew-S-Rosen** [comment]
Thanks! That was my issue. I installed `tblite` from conda-forge on v0.1.0 but was using `main` rather than the `v0.1.0` tag for this repo. I appreciate the help! The import works fine after making those consistent.

**Comment by awvwgk** [comment]
I will add a check to the extension module builder hook to catch the version mismatch (see https://github.com/tblite/tblite/pull/25).

**Comment by Andrew-S-Rosen** [comment]
Thanks! And just so I understand, in future `conda` versions will there still be a need to still `pip install` from the repo if I want to use the ASE calculator?

**Comment by awvwgk** [comment]
I actually want to create a `tblite-python` package which provides the Python bindings with conda as well at some point. Due to the early stage of the project I didn't submit a recipe yet, but I might do it for the version 0.2.0 release.

**Comment by Andrew-S-Rosen** [comment]
I think that's a great idea. Looking forward to it!

**Comment by awvwgk** [comment]
I just updated the build process for the Python bindings, which should allow easier standalone builds with meson using *tblite* as subproject of itself, via somewhat hacky symlinking:

```
mkdir python/subprojects
ln -s $PWD python/subprojects/tblite
meson setup _build_py python
meson compile -C _build_py
```

There is also a project called `mesonpep517`, which seems to provide a PEP517 compliant build interface to turn meson based projects into Python wheels / sdists. I've started playing around with this one in [`s-dftd3`](https://github.com/awvwgk/simple-dftd3/blob/main/python/mesonpep517.toml) and was actually able to create self-contained wheels, however it was not good enough to make it the default yet.

Therefore, I'm still somewhat hesitant to leave *setuptools* behind for building and installing the Python bindings of my projects, especially the editable install which was not accounted for with PEP517 and is essential for development IMO.
